example/balance/balance_corner2.c

```diff
@@ -0,0 +1,160 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with p4est; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+#ifndef P4_TO_P8
+#include <p4est_bits.h>
+#include <p4est_extended.h>
+#include <p4est_vtk.h>
+#else
+#include <p8est_bits.h>
+#include <p8est_extended.h>
+#include <p8est_vtk.h>
+#endif
+
+typedef struct
+{
+  p4est_topidx_t      a;
+}
+user_data_t;
+
+typedef struct
+{
+  sc_MPI_Comm         mpicomm;
+  int                 mpisize;
+  int                 mpirank;
+}
+mpi_context_t;
+
+/* refinement level initialization */
+static int          refine_level = 0;
+
+/* connectivity initialization function */
+static void
+init_fn (p4est_t * p4est, p4est_topidx_t which_tree,
+         p4est_quadrant_t * quadrant)
+{
+  user_data_t        *data = (user_data_t *) quadrant->p.user_data;
+
+  data->a = which_tree;
+}
+
+/* refinement function */
+static int
+refine_fn (p4est_t * p4est, p4est_topidx_t which_tree,
+           p4est_quadrant_t * quadrant)
+{
+  if ((int) quadrant->level >= (refine_level - (int) (which_tree % 3))) {
+    return 0;
+  }
+  if (quadrant->level == 1 && p4est_quadrant_child_id (quadrant) == 3) {
+    return 1;
+  }
+  if (quadrant->x == P4EST_LAST_OFFSET (2) &&
+      quadrant->y == P4EST_LAST_OFFSET (2)) {
+    return 1;
+  }
+  if (quadrant->x >= P4EST_QUADRANT_LEN (2)) {
+    return 0;
+  }
+
+  return 1;
+}
+
+int
+main (int argc, char **argv)
+{
+  int                 mpiret;
+  int                 wrongusage;
+  const char         *usage;
+  p4est_connectivity_t *connectivity;
+  mpi_context_t       mpi_context, *mpi = &mpi_context;
+  p4est_geometry_t   *geom;
+  p4est_t            *p4est;
+
+  /* initialize MPI and p4est internals */
+  mpiret = sc_MPI_Init (&argc, &argv);
+  SC_CHECK_MPI (mpiret);
+  mpi->mpicomm = sc_MPI_COMM_WORLD;
+  mpiret = sc_MPI_Comm_size (mpi->mpicomm, &mpi->mpisize);
+  SC_CHECK_MPI (mpiret);
+  mpiret = sc_MPI_Comm_rank (mpi->mpicomm, &mpi->mpirank);
+  SC_CHECK_MPI (mpiret);
+
+  sc_init (mpi->mpicomm, 1, 1, NULL, SC_LP_DEFAULT);
+  p4est_init (NULL, SC_LP_DEFAULT);
+
+  /* usage error if the input is not in the correct format */
+  usage =
+    "Arguments: <level>\n"
+    "   Level: controls the maximum depth of refinement\n";
+  wrongusage = 0;
+  if (!wrongusage && argc != 2) {
+    wrongusage = 1;
+  }
+  if (wrongusage) {
+    P4EST_GLOBAL_LERROR (usage);
+    sc_abort_collective ("Usage error");
+  }
+
+  /* assign variables based on configuration */
+  refine_level = atoi (argv[1]);
+
+  /* create connectivity and forest structures */
+  geom = NULL;
+#ifndef P4_TO_P8
+  connectivity = p4est_connectivity_new_bowtie ();
+  p4est = p4est_new_ext (mpi->mpicomm, connectivity, 15, 0, 0,
```

Comment:
The new_ext function is translated and can live outside the #ifdefs. We had historically chosen the parameter 15 for min_quadrants, but that makes the refinement partition-dependent, which we later found undesirable. This value can be set to 0. Conversely, the fill_uniform parameter would need to be set to 1 to make sure that the refinement looks the same independent of the number of cores used.